iscservice/demo/iscdemo.c

```diff
@@ -0,0 +1,607 @@
+		y->mr_idx++;
+		y->mr_nr++;
+	}
+
+	DBG2("xval=%lf yval=%lf\n", x->mr_val, y->mr_val);
+
+	if (ICT_MIN == op_type)
+		y->mr_val = min3(min_check(x->mr_val, y->mr_val), val1, val2);
+	else
+		y->mr_val = max3(max_check(x->mr_val, y->mr_val), val1, val2);
+
+	/* Update the resulting value index. */
+	if (y->mr_val == x->mr_val)
+		y->mr_idx = x->mr_idx;
+	else if (y->mr_val == val1)
+		y->mr_idx = x->mr_nr;
+	else if (y->mr_val == val2)
+		y->mr_idx = x->mr_nr + 1;
+
+	m0_free(buf);
+
+	return y;
+}
+
+enum elm_order {
+	ELM_FIRST,
+	ELM_LAST
+};
+
+static void set_idx(struct mm_result *res, enum elm_order e)
+{
+	if (ELM_FIRST == e)
+		res->mr_idx = 0;
+	else
+		res->mr_idx = res->mr_nr;
+}
+
+/**
+ * Process the first or the last values in the first or last results.
+ */
+static void check_edge_val(struct mm_result *res, enum elm_order e,
+			   enum isc_comp_type type)
+{
+	const char *buf;
+	double      val;
+
+	if (ELM_FIRST == e)
+		buf = res->mr_lbuf.b_addr;
+	else // last
+		buf = res->mr_rbuf.b_addr;
+
+	if (sscanf(buf, "%lf", &val) < 1) {
+		fprintf(stderr, "failed to parse egde value=%s\n", buf);
+		return;
+	}
+
+	DBG2("edge val=%lf (%s)\n", val, (ELM_FIRST == e) ? "first" : "last");
+
+	if (res->mr_nr == 1) {
+		res->mr_val = val;
+		return;
+	}
+
+	if (ICT_MIN == type && val < res->mr_val) {
+		res->mr_val = val;
+		set_idx(res, e);
+	} else if (ICT_MAX == type && val > res->mr_val) {
+		res->mr_val = val;
+		set_idx(res, e);
+	}
+}
+
+static void mm_result_free_xcode_bufs(struct mm_result *r)
+{
+	m0_free(r->mr_lbuf.b_addr);
+	m0_free(r->mr_rbuf.b_addr);
+}
+
+static void *minmax_output_prepare(struct m0_buf *result,
+				   bool last_unit,
+				   struct mm_result *prev,
+				   enum isc_comp_type type)
+{
+	int               rc;
+	struct mm_result  new = {};
+
+	rc = m0_xcode_obj_dec_from_buf(&M0_XCODE_OBJ(mm_result_xc, &new),
+				       result->b_addr, result->b_nob);
+	if (rc != 0) {
+		fprintf(stderr, "failed to parse result: rc=%d\n", rc);
+		goto out;
+	}
+	if (prev == NULL) {
+		M0_ALLOC_PTR(prev);
+		check_edge_val(&new, ELM_FIRST, type);
+		*prev = new;
+		goto out;
+	}
+
+	if (last_unit)
+		check_edge_val(&new, ELM_LAST, type);
+
+	/* Copy the current resulting value. */
+	if (op_result(prev, &new, type)) {
+		mm_result_free_xcode_bufs(prev);
+		*prev = new;
+	}
+ out:
+	/* Print the result. */
+	if (last_unit && prev != NULL) {
+		printf("idx=%lu val=%lf\n", prev->mr_idx, prev->mr_val);
+		mm_result_free_xcode_bufs(prev);
+		m0_free(prev);
+		prev = NULL;
+	}
+
+	return prev;
+}
+
+/**
+ * Apart from processing the output this can deserialize the buffer into
+ * a structure relevant to the result of invoked computation.
+ * and return the same.
+ */
+static void* output_process(struct m0_buf *result, bool last,
+			    void *out, enum isc_comp_type type)
+{
+	switch (type) {
+	case ICT_PING:
+		printf ("Hello-%s @%s\n", (char*)result->b_addr, (char*)out);
+		memset(result->b_addr, 'a', result->b_nob);
```

Comment:
This is merely to poison the buffer, right?

Reply:
I believe so, yes. To make sure that we have something useful in the next result.